internal/controller/pkg/revision/hook.go

```diff
@@ -215,6 +228,128 @@ func (h *ConfigurationHooks) Post(context.Context, runtime.Object, v1.PackageRev
 	return nil
 }
 
+// FunctionHooks performs operations for a function package that requires a
+// controller before and after the revision establishes objects.
+type FunctionHooks struct {
+	client         resource.ClientApplicator
+	namespace      string
+	serviceAccount string
+}
+
+// NewFunctionHooks creates a new FunctionHooks.
+func NewFunctionHooks(client resource.ClientApplicator, namespace, serviceAccount string) *FunctionHooks {
+	return &FunctionHooks{
+		client:         client,
+		namespace:      namespace,
+		serviceAccount: serviceAccount,
+	}
+}
+
+// Pre cleans up a packaged controller and service account if the revision is
+// inactive.
+func (h *FunctionHooks) Pre(ctx context.Context, pkg runtime.Object, pr v1.PackageRevision) error {
+	fo, _ := xpkg.TryConvert(pkg, &pkgmetav1alpha1.Function{})
+	pkgFunction, ok := fo.(*pkgmetav1alpha1.Function)
+	if !ok {
+		return errors.New(errNotFunction)
+	}
+
+	// TODO(ezgidemirel): update any status fields relevant to package revisions.
+
+	// Do not clean up SA and controller if revision is not inactive.
+	if pr.GetDesiredState() != v1.PackageRevisionInactive {
+		return nil
+	}
+
+	// NOTE(hasheddan): we avoid fetching pull secrets and controller config as
+	// they aren't needed to delete Deployment, ServiceAccount, and Service.
+	s, d, _, _ := buildFunctionDeployment(pkgFunction, pr, nil, h.namespace, []corev1.LocalObjectReference{})
+	if err := h.client.Delete(ctx, d); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteFunctionDeployment)
+	}
+	if err := h.client.Delete(ctx, s); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteFunctionSA)
+	}
+
+	return nil
+}
+
+// Post creates a packaged function deployment, service account, service and secrets if the revision is active.
+func (h *FunctionHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.PackageRevision) error { //nolint:gocyclo // See below
+	// TODO(ezgidemirel): Can this be refactored for less complexity?
+	po, _ := xpkg.TryConvert(pkg, &pkgmetav1alpha1.Function{})
+	pkgFunction, ok := po.(*pkgmetav1alpha1.Function)
+	if !ok {
+		return errors.New(errNotFunction)
+	}
+	if pr.GetDesiredState() != v1.PackageRevisionActive {
+		return nil
+	}
+	cc, err := h.getControllerConfig(ctx, pr)
+	if err != nil {
+		return err
+	}
+	ps, err := h.getSAPullSecrets(ctx)
+	if err != nil {
+		return err
+	}
+	s, d, svc, secSer := buildFunctionDeployment(pkgFunction, pr, cc, h.namespace, append(pr.GetPackagePullSecrets(), ps...))
+	if err := h.client.Apply(ctx, s); err != nil {
+		return errors.Wrap(err, errApplyFunctionSA)
+	}
+	owner := pr.GetOwnerReferences()
+	if err := h.client.Apply(ctx, secSer); err != nil {
+		return errors.Wrap(err, errApplyFunctionSecret)
+	}
+	if err := h.client.Apply(ctx, d); err != nil {
+		return errors.Wrap(err, errApplyFunctionDeployment)
+	}
+	if pr.GetTLSServerSecretName() != nil && pr.GetTLSClientSecretName() != nil {
```

Comment:
Why is the tls client secret populated for function revisions? it's not actually created.

Reply:
We're calling `GenerateServerCertificate` for functions and not generation client certificates. Since `NewTLSCertificateGenerator` gets client secret name, I added a nil check there.

Reply:
`NewTLSCertificateGenerator` should probably not require the name of both, but rather get some options, so that we don't have to fake it. I'd rather also not have `spec.webhookTLSSecretName`, nor `spec.tlsClientSecretName`, set if these are not actually used here, as that's misleading.

Reply:
made them option and removed the nil checks